### DISCUSSION TITLE
SOCKS5 proxies

### DISCUSSION THREAD

**Discussion by xamgore** [sindresorhus/got#1386] [Q&A] [answered]
Is it possible to exploit got.js with socks5 (4, 4a) proxies?

In [request](https://github.com/request/request) one could pass a custom `http.Agent` like [`socks5-http-client`](https://github.com/mattcg/socks5-http-client), that [redefines `createConnection` method](https://github.com/mattcg/socks5-http-client/blob/master/lib/Agent.js#L19) to return a socks5 socket (which is a `stream.Duplex`). A similar technique is used with [https-client](https://github.com/mattcg/socks5-https-client/blob/master/lib/Agent.js#L18).

With http2 there is a similar method [`createConnection`](https://nodejs.org/api/http2.html#http2_http2_connect_authority_options_listener), but I can't see a direct connection between it and got.js.

**Comment by sindresorhus** [comment]
You can pass a custom agent to Got too. It's in the docs.

**Comment by xamgore** [comment, marked as answer]
Got it working with http(s).

```js
  import { SocksProxyAgent } from 'socks-proxy-agent'

  await got.get('https://example.com', {
    http2: false,
    agent: {
      http: new SocksProxyAgent('socks://127.0.0.1:8080'),
      https: new SocksProxyAgent('socks://127.0.0.1:8080'),
      // http2: new SocksProxyAgent('socks://127.0.0.1:8080'),
    }
  })
```
